### ISSUE TITLE
Language server broken with clangd

### ISSUE THREAD

**Issue by rozhuk-im** [eranif/codelite#2499]
Then ```Language server``` is enabled it shows that it can not find some includes: project can be build by CodeLite internal build system, but these files in other project that is in workspace.
This include marked with red arrow like error and tooltip say that ```file not found```.
But then I click to ```Open include file``` it opens.

This also causes that ```Goto Declaration``` and ```Goto Implementation``` not work - nothing happen.

I suspect there is 2 issues:
1. ```Goto Declaration``` and ```Goto Implementation``` must work even if ```Language server``` return error.
2. ```Language server``` does not pass all includes path to ```clangd``` and it fails.

As work around I disable ```Language server``` to allow code navigation.

**Comment by eranif** [comment]
I will need some more details:
- Is it a project build with a custom `Makefile` or built using CodeLite default build system?
- OS?

You can easily see which include paths and macros are passed to `clangd` by finding the `compile_flags.txt` file on the file system.

**Comment by rozhuk-im** [comment]
It is static lib, build with CodeLite build system.
FreeBSD 12.1

In project folder ```compile_flags.txt``` contains:
```-I/app/codelite/xxxldap/$(WorkspacePath)/src/lib/xxxutil```
but file in: ```/app/src/lib/xxxutil```

**Comment by eranif** [comment]
I wonder why do you have `$(WorkspacePath)` in your `compile_flags.txt` ?
It's not something that should appear in that file...

So the bug is that CodeLite does not expand `$(WorkspacePath)` as it should before creating the `compile_flags.txt` file.

As a temporary workaround, you can use relative paths instead of this macro

**Comment by rozhuk-im** [comment]
There is also ```/app/codelite/xxxldap/``` must not be included.
In ```/app/codelite/``` we store only codelite things, all code in ```/app/src/...```.

**Comment by eranif** [comment]
CodeLite generates `compile_flags.txt` *per* project.
i.e. you should have multiple `compile_flags.txt` files - one per project folder. 
Note that if all your projects are located under the same directory, then it is a problem.

**Comment by eranif** [comment]
This commit should fix the macro expansion bug:
https://github.com/eranif/codelite/commit/a648b4e655f5f6592fcb5ac30bb8cfe06f361e36

**Comment by rozhuk-im** [comment]
This fix project ```compile_flags.txt``` generation but nothing changes.

**Comment by eranif** [comment]
so what is the content of the `compile_flags.txt` under the same folder that the code completion does not work for you?. Does it make sense to you (the content of that file?)

**Comment by rozhuk-im** [comment]
```compile_flags.txt``` file is ok.
But looks like it does not passed to clangd.

I get all lines from ```compile_flags.txt``` and pass it to ```cc``` with .c file name and got only ```undefined symbol``` errors - it compile ok and fail to link (as expected).

To be clear:
```/app/codelite/xxxldap/``` - project, ```compile_flags.txt``` and other CodeLite staff is here
```/app/src/lib/xxxldap/``` - *.c, *.h files from this project.

**Comment by eranif** [comment]
There is no need to "pass" it to `clangd`. If this file is located under the same folder as the file that you currently trying to code complete, `clangd` *should* pick it up

Which version of clangd are you using?
Did you try to restart `clangd` -> from within CodeLite `Ctrl-Shift-P` , and type `restart lang` and hit ENTER

**Comment by rozhuk-im** [comment]
My *.c files here: ```/app/src/lib/xxxldap/```
CodeLite write ```compile_flags.txt``` to ```/app/codelite/xxxldap/``` - **this is different folder**, not project folder.
I copy ```compile_flags.txt``` to ```/app/codelite/xxxldap/``` and this helps:
1. Code navigation from this file work.
2. No error markers shown.


Please, fix it in both places:

1. ```Goto Declaration``` and ```Goto Implementation``` menu items must work in any cases, not depend on ```Language server```

2. Store ```compile_flags.txt``` files to folders with *.c files - it may be stored in many folders outside project dir.

**Comment by eranif** [comment]
1. I would create a symlink on the file system, so next time this file is updated, you have updated one
2. So basically, your project file (`.project`) and source files are on different trees?
3. Goto * -> the built in code completion is going away soon. So any fix will need to be implemented to the language server. The current C++ language is evolving too fast to keep up with it, and the current built-in parser is stack somewhere in the early 2000's. The only way forward is with LSP
4. Yes, I figured the problem (I mentioned that in my point #2): your source files are kept in a different source tree (not even the same parent folder)

**Comment by rozhuk-im** [comment]
1. Thanks. It is ok for me, hope windows users have symlinks too :)

2. Yes, as you can see :)
Also I have projects (like liblcb) where one project file and multiple dirs with *.c and *.h files.

3. I suspect that ```Language server``` never work before on my system and some how auto activated on last updates. I check old CodeLite - they ```Goto Implementation``` broken too if ```Language server``` is active.
Auto completion is important but code navigation is very important for me :)

**Comment by lwchkg** [comment]
+1, See below for a list of errors, and it's really broken.

I'm not sure what is happening, but looks like `compile_commands.json` is not generated correctly.  I do see a `compile_commands.json` in `.codelite`, but it appeared to be several years old because it points to a compiler I've uninstalled long long before.

![codelite_7DgSoBHeBv](https://user-images.githubusercontent.com/4113719/81958524-b8292900-9640-11ea-8e0c-ad666ddda84b.png)

**Comment by eranif** [comment]
@lwchkg is this file part of a project?
If so, what kind of a workspace dd you open in CodeLite? and what OS is this?

**Comment by eranif** [comment]
@rozhuk-im FYI: there is an option in CodeLite to generate `compile_commands.json` file (it's different than the `compile_flags.txt` file)
I did not mention this until now, because it was broken (in a way):
It did generate `compile_commands.json` file, however, it was an addition to the `compile_flags.txt` files. The problem is that `clangd` gives `compile_flags.txt` a higher priority, so it was useless.

I fixed this in CodeLite: i.e. when CodeLite is asked to create `compile_commands.json` it will no longer generate the `compile_flags.txt` files.

How does this fix your issue?
Well, `compile_commands.json` file contains instructions **per file** on how to build it. so it does not matter where this file is placed, `clangd` knows the exact command for compiling this specific file.

To use this:

- Build latest CodeLite
- find and delete all instances of `compile_flags.txt` that you have in your workspace
- Issue a build within CodeLite - at the end of the build, CodeLite will generate this file and it will restart `clangd` to load this file

**Comment by lwchkg** [comment]
@eranif  Works for me today, no more errors.  Errors happened BEFORE I added MINGW to system path, and disappeared after that.  I've added that because programs compiled with MINGW can't run in CMD otherwise.

Anyway, I use a C++ workspace, with a lot of single-file projects, mainly used for Google Code Jam. The screenshot you see is one of them.  (PS: I just lost in Round 2, LOL.)  The code compiles with zero warnings, so the output shouldn't be useful.

OS: Windows 10.
